File: scripts/process_vehicles.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

// Turkish alphabet for sorting (case-insensitive mapping)
const TR_ALPHABET = "abcçdefgğhıijklmnoöprsştuüvyz";
const TR_CHARS = {
    'Ç': 'ç', 'Ğ': 'ğ', 'I': 'ı', 'İ': 'i', 'Ö': 'ö', 'Ş': 'ş', 'Ü': 'ü',
    'C': 'c', 'G': 'g', 'O': 'o', 'S': 's', 'U': 'u', 'I': 'i'
};

function trLowercase(str) {
    if (!str) return '';
    return str.toString().replace(/[ABCÇDEFGĞHIİJKLMNOÖPRSŞTUÜVYZ]/g, (m) => {
        const mapped = TR_CHARS[m];
        return mapped || m.toLowerCase();
    }).toLowerCase();
}

function trSortKey(s) {
    if (typeof s !== 'string') return s;
    const lower = trLowercase(s);
    return Array.from(lower).map(c => {
        const idx = TR_ALPHABET.indexOf(c);
        return idx !== -1 ? idx.toString().padStart(2, '0') : c;
    }).join('');
}

function processVehicles() {
    const inputFile = path.join(__dirname, '202604R2.xlsx');
    const outputFile = path.join(process.cwd(), 'vehicles.json');

    console.log(`Reading ${inputFile}...`);
    try {
        const workbook = XLSX.readFile(inputFile);
        const sheetName = workbook.SheetNames[0];
        const worksheet = workbook.Sheets[sheetName];
        const data = XLSX.utils.sheet_to_json(worksheet, { defval: 0 });

        console.log(`Processing ${data.length} rows...`);

        const brandCol = "Marka Adı";
        const modelCol = "Tip Adı";
        const years = Array.from({ length: 2026 - 2012 + 1 }, (_, i) => (2026 - i).toString());

        // Grouping
        const brandsMap = new Map();

        data.forEach(row => {
            const brand = (row[brandCol] || '').toString().trim();
            const model = (row[modelCol] || '').toString().trim();

            if (!brand || !model) return;

            if (!brandsMap.has(brand)) {
                brandsMap.set(brand, new Map());
            }

            const modelsMap = brandsMap.get(brand);
            if (!modelsMap.has(model)) {
                modelsMap.set(model, new Set());
            }

            const activeYears = modelsMap.get(model);
            years.forEach(year => {
                // xlsx usually reads headers as they are. Check both string and potentially numeric if it was converted
                const val = row[year];
                if (val > 0) {
                    activeYears.add(parseInt(year));
                }
            });
        });

        // Convert Map to Array and Sort
        const result = Array.from(brandsMap.entries())
            .map(([brand, modelsMap]) => {
                const models = Array.from(modelsMap.entries())
                    .map(([model, activeYearsSet]) => {
                        return {
                            package_name: model,
                            active_years: Array.from(activeYearsSet).sort((a, b) => b - a)
                        };
                    })
                    .filter(m => m.active_years.length > 0)
                    .sort((a, b) => trSortKey(a.package_name).localeCompare(trSortKey(b.package_name)));

                return {
                    brand: brand,
                    models: models
                };
            })
            .filter(b => b.models.length > 0)
            .sort((a, b) => trSortKey(a.brand).localeCompare(trSortKey(b.brand)));

        console.log(`Writing output to ${outputFile}...`);
        fs.writeFileSync(outputFile, JSON.stringify(result, null, 2), 'utf-8');
        console.log("Successfully created vehicles.json!");

    } catch (error) {
        console.error("Error processing file:", error);
    }
}

processVehicles();

File: scripts/process_vehicles.py
import pandas as pd
import json
import os

# Turkish alphabet for sorting
TR_ALPHABET = "abcçdefgğhıijklmnoöprsştuüvyz"
TR_LOW_MAP = str.maketrans("ABCÇDEFGĞHIİJKLMNOÖPRSŞTUÜVYZ", "abcçdefgğhıijklmnoöprsştuüvyz")

def tr_sort_key(s):
    """Sort key for Turkish strings."""
    if not isinstance(s, str):
        return s
    return [TR_ALPHABET.index(c) if c in TR_ALPHABET else ord(c) for c in s.translate(TR_LOW_MAP)]

def process_vehicles():
    input_file = "scripts/202604R2.xlsx"
    output_file = "vehicles.json"
    
    if not os.path.exists(input_file):
        print(f"Error: {input_file} not found.")
        return

    print(f"Reading {input_file}...")
    try:
        # We use openpyxl engine as requested/pre-installed
        df = pd.read_excel(input_file, engine='openpyxl')
    except Exception as e:
        print(f"Error reading Excel: {e}")
        return

    print("Processing columns...")
    # Standardize column names
    df.columns = [str(c).strip() for c in df.columns]
    
    brand_col = "Marka Adı"
    model_col = "Tip Adı"
    
    # Year columns 2026 to 2012
    year_cols = [str(y) for y in range(2026, 2011, -1)]
    
    # Verify columns exist
    missing = [c for c in [brand_col, model_col] + year_cols if c not in df.columns]
    if missing:
        print(f"Warning: Missing columns: {missing}")
        # Try to find numeric versions of years if strings match
        for m in missing:
            if m.isdigit():
                if int(m) in df.columns:
                    df.rename(columns={int(m): m}, inplace=True)
                elif float(m) in df.columns:
                    df.rename(columns={float(m): m}, inplace=True)

    # Re-verify
    missing = [c for c in [brand_col, model_col] + year_cols if c not in df.columns]
    if missing:
        print(f"Critical Error: Still missing columns: {missing}")
        return

    result = []
    
    # Group by Brand
    brands = df[brand_col].unique()
    # Sort brands
    sorted_brands = sorted(brands, key=tr_sort_key)
    
    for brand in sorted_brands:
        if pd.isna(brand): continue
        
        brand_df = df[df[brand_col] == brand]
        models_list = []
        
        # Group by Model
        models = brand_df[model_col].unique()
        sorted_models = sorted(models, key=tr_sort_key)
        
        for model in sorted_models:
            if pd.isna(model): continue
            
            model_df = brand_df[brand_df[model_col] == model]
            
            # Combine multiple rows for the same model if they exist (though usually unique per model in taxonomy)
            active_years = set()
            for year in year_cols:
                if (model_df[year] > 0).any():
                    active_years.add(int(year))
            
            if active_years:
                models_list.append({
                    "package_name": str(model).strip(),
                    "active_years": sorted(list(active_years), reverse=True)
                })
        
        if models_list:
            result.append({
                "brand": str(brand).strip(),
                "models": models_list
            })

    print(f"Writing output to {output_file}...")
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(result, f, ensure_ascii=False, indent=2)
    
    print("Successfully created vehicles.json!")

if __name__ == "__main__":
    process_vehicles()
